src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -59,6 +60,9 @@ def __init__(
         self._make_install_req_from_spec = make_install_req
         self._force_reinstall = force_reinstall
         self._ignore_requires_python = ignore_requires_python
+        self._upgrade_strategy = upgrade_strategy
+
+        self.root_reqs = set()  # type: Set[str]
```

Comment:
Having this as an instance variable means the factory would have a different state before and after performing the resolution. I know we only resolve once in its lifetime, but this still makes uneasy because the class is otherwise implemented as stateless (except the candidate cache, which I can get behind of).

Reply:
I don't see how to avoid that. We can't check `is_direct` when we're looking at the candidate, because it's a *project name* that's eligible for upgrade (and so recorded in `root_reqs`) not the specific requirement (consider `pip install A B` where B depends on `A>1.0`. The candidates for the `A>1.0` requirement must be ordered as eligible for upgrade, even though that requirement isn't direct.

I'd be OK with adding a `Factory.reset` method, and calling it from `Resolver.resolve` - but it feels like it's taking caution a bit far. How about for now, I add an assertion in `Resolver.resolve` to check that `factory.root_reqs` is empty at the start of the resolve, then if we later find we violate that assertion, we can fix the issue then?

Reply:
I understand (I don’t have anything to offer myself). Let's get this in first then, refactoring like this can always wait until it actually matters.